core/state_transition.go

```diff
@@ -197,6 +199,9 @@ func (st *StateTransition) to() common.Address {
 func (st *StateTransition) buyGas() error {
 	mgval := new(big.Int).SetUint64(st.msg.Gas())
 	mgval = mgval.Mul(mgval, st.gasPrice)
+	if st.msg.L1Cost() != nil {
+		mgval = mgval.Add(mgval, st.msg.L1Cost())
+	}
 	balanceCheck := mgval
 	if st.gasFeeCap != nil {
 		balanceCheck = new(big.Int).SetUint64(st.msg.Gas())
```

Comment:
Sorry that I missed this, but you also need to add the `L1Cost` to the balance check.
`st.gasFeeCap != nil => 1559 is enabled` and when 1559 is enabled, you must be able to pay for the max fee that you asked for rather than the actual fee you paid. Pre 1559 the fee you pay is the same as the max fee you put so it takes a different path.